### PR TITLE
Onyx Connection Manager

### PR DESCRIPTION
**Pull request by fabioh8010**
### Details

This PR implements a general connection manager to handle all Onyx connections. Key features:

* **Connection reuse:** Connections made to the same Onyx key with the same configurations are reused instead of being kept separately, thus reducing JS processing and overhead when the Onyx key's data changes and needs to broadcast to the connections. This approach can significantly reduce the total number of connections during the application's runtime and it's supported by both `Onyx.connect()` and `useOnyx()` subscribers. Since `withOnyx()` subscribers connects to Onyx keys in a different way than the first ones, they can't be reused and will work the same way as before.
* **Centralized logic:** Both `Onyx.connect()`, `useOnyx()` and `withOnyx()` subscribes will use the connection manager to handle their connections, keeping the logic centralized and allowing easier maintenance and future improvements.

Slack thread: https://expensify.slack.com/archives/C05LX9D6E07/p1723058474781739

E/App PR: https://github.com/Expensify/App/pull/44987

### Related Issues

https://github.com/Expensify/App/issues/47143

### Automated Tests

* Tests related to Onyx Connection Manager.
* Additional tests to `useOnyx`.
* Adjusting other tests.

### Manual Tests

Common actions in the App, e.g.:

1. Open a chat.
2. Send a message, react to that message.
3. Reply to that message as a thread.
4. Submit an expense, open that expense.
5. Assign a task, open that task and mark it as complete.
6. Go to Settings and change the App theme.

### Author Checklist

- [x] I linked the correct issue in the `### Related Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/f98623d4-f100-4de1-b68d-9501c1f8f9d2

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/739773f7-80d7-42f3-830c-84f0aa7e308f

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/2bb88f52-9899-4146-a6b6-9d48f84b9edf

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/cc357b80-a4fa-475f-bdab-303bc9c688d3

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/76775ece-1043-46d3-91cc-4a23b3435aa8

https://github.com/user-attachments/assets/6c476720-c54d-4667-8857-e471fe7f44f4

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/48d233ab-4cc0-49d9-b0e1-8e13ccd5f570

</details>

### CODE REVIEW

File: lib/Str.ts
Comment (tgolen) [inline lib/Str.ts:29]: `prefix` is not used. Can you remove it, please?

File: lib/OnyxUtils.ts
Comment (tgolen) [inline lib/OnyxUtils.ts:1348]: Maybe this should throw instead?

**Review by tgolen — CHANGES_REQUESTED** [review]

File: lib/OnyxUtils.ts
Comment (fabioh8010) [reply lib/OnyxUtils.ts:1348]: Hmm this log was already in this function before my PR -- maybe we could remove it but I would like to do this in another PR to reduce the chance of regressions. WDYT?

**Comment by mountiny** [comment]
@fabioh8010 can you apply the changes from here to the App PR https://github.com/Expensify/App/pull/44987 (or create a separate draft for testing) so we can create a build and ask QA for testing before merging this?

**Comment by mountiny** [comment]
Asked for a C+ volunteer

File: tests/unit/onyxCacheTest.tsx
Comment (mountiny) [inline tests/unit/onyxCacheTest.tsx:551]: Why this change?

**Review by mountiny — COMMENTED** [review]
Havent done a full review yet, but excited for this change

File: lib/OnyxUtils.ts
Comment (tgolen) [reply lib/OnyxUtils.ts:1348]: Sure, that sounds fine with me.

File: lib/OnyxConnectionManager.ts
Comment (blazejkustra) [inline lib/OnyxConnectionManager.ts:144]: I may not understand this completely, so feel free to correct me. A `callback` is defined here, that in this exact line calls all callbacks for a given `connectionID`, wouldn't it fire callbacks indefinitely? (first `callback` calls all `callbacks` that call all `callbacks` again and again 😅)

**Review by blazejkustra — CHANGES_REQUESTED** [review]

File: lib/OnyxConnectionManager.ts
Comment (chrispader) [reply lib/OnyxConnectionManager.ts:144]: I think there is a difference between the callback of a `subscription` and a `connection`, right?

Here the callbacks of the `connection` are triggered, but this will not trigger the (outer) callback of the `subscription`.

File: tests/unit/onyxCacheTest.tsx
Comment (fabioh8010) [reply tests/unit/onyxCacheTest.tsx:551]: We generate 10 connections from 0 to 9, and this line is connecting again to the connection 9. Since the connection manager is in place this won't create a new connection, but rather reuse the connection 9 previously created. This part of the test requires a new connection to be made, so I just incremented the number. Another way to solve is:

```ts
Onyx.connect({key: `${ONYX_KEYS.COLLECTION.MOCK_COLLECTION}9`, callback: jest.fn(), reuseConnection: false});
```

File: lib/OnyxUtils.ts
Comment (fabioh8010) [reply lib/OnyxUtils.ts:1352]: Actually it has to be `// by calling OnyxUtils.unsubscribeFromKey(subscriptionID).` 😅

File: lib/OnyxConnectionManager.ts
Comment (fabioh8010) [reply lib/OnyxConnectionManager.ts:79]: We use binding on `OnyxCache` too, I think for safety purposes it's better to leave it, I can add a comment.

**Review by chrispader — CHANGES_REQUESTED** [review]
I really like this PR. Awesome work!

I just have a few questions, but the code seems very logical and well designed.

File: lib/OnyxConnectionManager.ts
Comment (fabioh8010) [reply lib/OnyxConnectionManager.ts:8]: Because `key` is defined as `TKey` on `DefaultConnectCallback` type, while `CollectionConnectCallback` has `key` as `undefined`, which creates some typing problems here.

File: lib/OnyxConnectionManager.ts
Comment (fabioh8010) [reply lib/OnyxConnectionManager.ts:38]: `OnyxValue` already includes `OnyxCollection`, I put this way to avoid other typing problems would arise.

File: lib/OnyxConnectionManager.ts
Comment (fabioh8010) [reply lib/OnyxConnectionManager.ts:141]: I'm going to test and let you know!

File: lib/OnyxConnectionManager.ts
Comment (chrispader) [reply lib/OnyxConnectionManager.ts:38]: This can also be a single Onyx entry, right? So this mustn't be a collection

File: lib/OnyxConnectionManager.ts
Comment (chrispader) [inline lib/OnyxConnectionManager.ts:48]: Same as above

File: lib/OnyxConnectionManager.ts
Comment (chrispader) [reply lib/OnyxConnectionManager.ts:8]: Same thought. Was this to simplify casting somewhere?

File: lib/OnyxUtils.ts
Comment (fabioh8010) [reply lib/OnyxUtils.ts:1337]: I ended up with this for now

```ts
function hasWithOnyxInstance<TKey extends OnyxKey>(mapping: ConnectOptions<TKey>) {
    return 'withOnyxInstance' in mapping && mapping.withOnyxInstance;
}
```

I tried type guarding but it created TS errors on `OnyxUtils`. I will try to go back to it later to see if I can find a solution

File: lib/OnyxConnectionManager.ts
Comment (fabioh8010) [reply lib/OnyxConnectionManager.ts:38]: `OnyxValue` can be both OnyxEntry and OnyxCollection!

File: lib/OnyxConnectionManager.ts
Comment (chrispader) [reply lib/OnyxConnectionManager.ts:141]: I don't think it's redundant here and we cannot easily change this to only use the `OnyxCache`.

It makes sense to keep the caching logic in `keyChanged`/`keysChanged`. We receive cached collections with `getCachedCollection(collectionKey)` and then build the updated collection in `keysChanged`, whereas for single key updates the value is just passed directly to `keyChanged`.

In the `OnyxConnectionManager` we don't want to differentiate connections between collection and single key, it should only be concerned about handling the the communication between `subscription <> connection`. Therefore, i think it makes sense to keep the cache for value and key in the `connectionsMap`.

Wdyt @fabioh8010 ?

File: lib/OnyxConnectionManager.ts
Comment (chrispader) [reply lib/OnyxConnectionManager.ts:79]: Agree, we should keep it.

File: lib/OnyxConnectionManager.ts
Comment (fabioh8010) [reply lib/OnyxConnectionManager.ts:144]: Yes, the callbacks that are being fired are the ones supplied when the consumer call `Onyx.connect` (or inside `useOnyx`), while this `callback` defined here is related to the connection, that can be reused and fire multiple consumer's callbacks.

File: lib/OnyxUtils.ts
Comment (chrispader) [reply lib/OnyxUtils.ts:1337]: Sounds good. I think it should work though, because `WithOnyxConnectOptions` is just the base type of `Mapping`

File: lib/OnyxConnectionManager.ts
Comment (chrispader) [reply lib/OnyxConnectionManager.ts:8]: > Because `key` is defined as `TKey` on `DefaultConnectCallback` type, while `CollectionConnectCallback` has `key` as `undefined`, which creates some typing problems here.

This brings me to [this comment](https://github.com/Expensify/react-native-onyx/pull/567#discussion_r1723116323) again... Why is it `undefined` in the first place?

**Review by chrispader — CHANGES_REQUESTED** [review]

File: lib/OnyxUtils.ts
Comment (fabioh8010) [reply lib/OnyxUtils.ts:1267]: Good catch, makes sense!

File: lib/OnyxUtils.ts
Comment (fabioh8010) [reply lib/OnyxUtils.ts:634]: Not sure why, it was always like this before – however, I can make this change to pass the collection key

File: lib/types.ts
Comment (fabioh8010) [reply lib/types.ts:280]: Not sure why, it was always like this before – however, I can make this change to pass the collection key

File: lib/OnyxConnectionManager.ts
Comment (fabioh8010) [reply lib/OnyxConnectionManager.ts:37]: Actually it's correct, the callback I pass to `OnyxUtils.subscribeToKey` is the one I use to cache these values.

File: lib/OnyxConnectionManager.ts
Comment (fabioh8010) [reply lib/OnyxConnectionManager.ts:8]: > Same thought. Was this to simplify casting somewhere?

Yes, exactly.

> This brings me to https://github.com/Expensify/react-native-onyx/pull/567#discussion_r1723116323 again... Why is it undefined in the first place?

Answered [here](https://github.com/Expensify/react-native-onyx/pull/567#discussion_r1723366610). But even with that changes I still need this type, using `DefaultConnectCallback` and `CollectionConnectCallback` leaves to type incompatibilities

File: lib/OnyxConnectionManager.ts
Comment (fabioh8010) [reply lib/OnyxConnectionManager.ts:141]: Yes I agree, and it shouldn't be a problem regarding to performance as we are just storing references (no deep copying).

File: lib/OnyxUtils.ts
Comment (chrispader) [reply lib/OnyxUtils.ts:634]: True, we never passed the key to collection callbacks. But now that we unified the signature type of all callbacks, i think it makes sense to pass the collection key rather than just `undefined`.

File: lib/OnyxConnectionManager.ts
Comment (chrispader) [reply lib/OnyxConnectionManager.ts:37]: Ah misread the comment, nevermind sorry 🙌🏼

File: lib/OnyxConnectionManager.ts
Comment (chrispader) [reply lib/OnyxConnectionManager.ts:48]: I think this comment got shifted to the wrong line... same as this: https://github.com/Expensify/react-native-onyx/pull/567#discussion_r1723128171

my bad sorry

File: API.md
Comment (fabioh8010) [reply API.md:79]: Added them back.

**Review by blazejkustra — APPROVED** [review]
Great improvement @fabioh8010! 🚀

**Review by tgolen — APPROVED** [review]

**Comment by tgolen** [comment]
Now that this is approved, let's talk about rolling this out. Any thoughts on how to do it safely?

Normally, I would say that QA should do a full regression test on an ad-hoc build of E/App. However, they currently have 4-5 ad-hoc builds they are doing regression tests on, so it could be a while before they get to this.

I'm sort of thinking that the nature of this PR means that either everything works as normal, or everything will be massively broken 😂 So, maybe we can get away with:

- HOLD this PR so it is not merged yet
- Create an E/App PR that points package.json to the commit of this branch
- Test it out and verify there are no issues with most of our core flows (maybe have a C+ or two do some testing)
- Once the E/App PR is tested and any bugs are fixed, merge this PR
- Update the E/App PR to point package.json to the newly released version

**Comment by fabioh8010** [comment]
Hi @tgolen , I already have the [E/App PR here](https://github.com/Expensify/App/pull/44987) pointing to a release I made with these changes in my repo.

The C+ assigned [tested and didn't found any bugs so far](https://github.com/Expensify/App/pull/44987#issuecomment-2299987207).

I think the best approach would be having the full regression test with ad-hoc build, even if we have to wait a little bit. In this way we would ensure more safety before these changes are merged. WDYT?

**Comment by tgolen** [comment]
Ah, nice. You were well ahead of me!

I discussed with @mountiny @roryabraham @luacmartins and @AndrewGable and we all agreed that this doesn't need ad-hoc testing. That's mostly based on the idea that fundamentally, this PR will either work fine, or break everything. It is also a simple change to revert (for E/App at least). Because of that, we're fine with it just going to staging like normal.

Is that OK?

**Review by mountiny — APPROVED** [review]
Agreed with @tgolen We could try this out in main and staging given @ikevin127 did not find any problems so far

**Comment by fabioh8010** [comment]
Yeah I think it makes sense, let's :shipit: then

**Pull request merged**
